### DISCUSSION TITLE
injecting-services-into-validationattributes-in-asp-net-core/

### DISCUSSION THREAD

**Discussion by andrewlock** [andrewlock/blog-comments#169] [Announcements]
# injecting-services-into-validationattributes-in-asp-net-core/

This post describes how to use dependency injection with ValidationAttributes in ASP.NET Core, and the process I took in trying to figure out how!

https://andrewlock.net/injecting-services-into-validationattributes-in-asp-net-core/

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at October 01 2016, 10:13, in reply to [Paul H](https://disqus.com/by/disqus_9VS90SRLfx/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/169#discussioncomment-6005498)</em>

---

<p>Yup, I completely agree, Fluent Validation certainly works well with more complex validation, especially when it's related to business concerns.</p><p>The project was already using DataAnnotations as it only needed simple form validation initially. In this case, the additional validation required was a Luhn check, so it wasn't complicated enough to warrant a switch from DataAnnotations, but we still wanted to be able to unit test it, hence the service.</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at October 01 2016, 10:16, in reply to [valdis iljuconoks](https://disqus.com/by/tech_fellow/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/169#discussioncomment-6005500)</em>

---

<p>I agree, FluentValidation is certainly more powerful than the attribute approach; being able to easily validate across multiple properties is a requirement that often comes up pretty quickly. However for very simple form validation, the declarative DataAnnotation attributes can be quite appealing.</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at October 01 2016, 10:18, in reply to [TM](https://disqus.com/by/disqus_CjezMYwXUO/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/169#discussioncomment-6005503)</em>

---

<p>Should it though, if the constructor dependencies prevent you actually being able to use it as an attribute? The docs imply not :p <a href="https://docs.asp.net/en/latest/mvc/controllers/filters.html#configuring-filters" rel="nofollow noopener" title="https://docs.asp.net/en/latest/mvc/controllers/filters.html#configuring-filters">https://docs.asp.net/en/latest/mvc/controllers/filters.html#configuring-filters</a></p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at October 07 2016, 09:47, in reply to [Richar Flores](https://disqus.com/by/richar_flores/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/169#discussioncomment-6005506)</em>

---

<p>That looks like a DI issue - I suggest you raise a question on stackoverflow. Cheers</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at April 25 2017, 08:41, in reply to [Mark](https://disqus.com/by/disqus_FGw5PbWHfC/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/169#discussioncomment-6005513)</em>

---

<p>No, just an oversight on my part - GetService returns an Object so you will need to cast it as you have. <br>I'll update the example, thanks :)</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at September 05 2019, 10:23, in reply to [Vedant](https://disqus.com/by/disqus_xRXR76Hr9g/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/169#discussioncomment-6005519)</em>

---

<p>If you need to, you can do something like the following: <br><code>new ServiceCollection().AddSingleton&lt;myservice&gt;().BuildServiceProvider()</code></p><p>However I would suggest you move all your validation logic into a separate service that uses standard dependency injection, so is easy to test. The ValidationAttribute then becomes very simple - it just retrieves the service and runs it - so no need to test that part.</p><p>Hope that helps!</p>

[a comment by an automated account is not shown]

**Comment by dlidstrom** [comment]
Requiring something from the outer world should not be uncommon or undesirable. For example, I can see a need to require looking up settings for use in validations (max days, or any other requirements you have). Fluent Validation is not very nice (IMO) as it moves the simple validation logic too far from its usage and it isn't easy to find. Similar issue as using AutoMapper.
